plugins/gdk-pixbuf/pixbufloader-jxl.c

```diff
@@ -3,258 +3,257 @@
+           decoder_state->parallel_runner)) != JXL_DEC_SUCCESS) {
     g_set_error(error, GDK_PIXBUF_ERROR, GDK_PIXBUF_ERROR_FAILED,
-                "Failed to create output pixbuf");
-    free(decoded);
-    return NULL;
-  }
-
-  return pixbuf;
-}
-
-static gpointer gdk_pixbuf__jxl_image_begin_load(
-    GdkPixbufModuleSizeFunc size_func, GdkPixbufModulePreparedFunc prepare_func,
-    GdkPixbufModuleUpdatedFunc update_func, gpointer user_data,
-    GError **error) {
-  JxlContext *context = g_new(JxlContext, 1);
-  context->size_func = size_func;
-  context->prepare_func = prepare_func;
-  context->update_func = update_func;
-  context->user_data = user_data;
-  context->error = error;
-
-  context->increment_buffer = open_memstream(&context->increment_buffer_ptr,
-                                             &context->increment_buffer_size);
-
-  if (!context->increment_buffer) {
-    perror("Cannot create increment buffer.");
-    g_free(context);
-    return NULL;
+                "JxlDecoderSetParallelRunner failed: %x", status);
+    goto cleanup;
   }
-
-  return context;
-}
-
-static gboolean gdk_pixbuf__jxl_image_stop_load(gpointer user_context,
-                                                GError **error) {
-  JxlContext *context = (JxlContext *)user_context;
-
-  int status = fflush(context->increment_buffer);
-  status |= fseek(context->increment_buffer, 0L, SEEK_SET);
-
-  if (status != 0) {
-    perror("Cannot flush and rewind increment buffer.");
-    fclose(context->increment_buffer);
-    free(context->increment_buffer_ptr);
-    g_free(context);
-    return FALSE;
+  if ((status = JxlDecoderSubscribeEvents(
+           decoder_state->decoder,
+           JXL_DEC_BASIC_INFO | JXL_DEC_COLOR_ENCODING | JXL_DEC_FULL_IMAGE)) !=
+      JXL_DEC_SUCCESS) {
+    g_set_error(error, GDK_PIXBUF_ERROR, GDK_PIXBUF_ERROR_FAILED,
+                "JxlDecoderSubscribeEvents failed: %x", status);
+    goto cleanup;
   }
 
-  context->pixbuf =
-      gdk_pixbuf__jxl_image_load(context->increment_buffer, error);
-
-  gint width = gdk_pixbuf_get_width(context->pixbuf);
-  gint height = gdk_pixbuf_get_height(context->pixbuf);
-  if (context->size_func) {
-    context->size_func(&width, &height, context->user_data);
-  }
+  decoder_state->pixel_format.data_type = JXL_TYPE_FLOAT;
+  decoder_state->pixel_format.endianness = JXL_NATIVE_ENDIAN;
 
-  if (context->prepare_func) {
-    (*context->prepare_func)(context->pixbuf, NULL, context->user_data);
-  }
+  return decoder_state;
+cleanup:
+  JxlResizableParallelRunnerDestroy(decoder_state->parallel_runner);
+  JxlDecoderDestroy(decoder_state->decoder);
+  return NULL;
+}
 
-  if (context->update_func) {
-    (*context->update_func)(
-        context->pixbuf, 0, 0, gdk_pixbuf_get_width(context->pixbuf),
-        gdk_pixbuf_get_height(context->pixbuf), context->user_data);
+static gboolean stop_load(gpointer context, GError **error) {
+  GdkPixbufJXLDecoderState *decoder_state = context;
```

Comment:
You should explicitly cast this here. Implicit casting of `void *` to anything else will fail if you compile this with a C++ compiler.

Reply:
Compiling this with a C++ compiler will do Very Bad Things, such as mangle the fill_info and fill_vtable functions - this will result in a non-usable plugin, so I prefer failing at compile time :)